### PR TITLE
perf: add `ETL` to `init_from_state_dump`

### PR DESCRIPTION
**Pull request by joshieDo**
Adds ETL to the init-state flow, and reverts `StateReverts::write_to_db` changes 

### Reasoning:
`reth init_state` will initialize the node with a genesis state.

There's the option to pass on a state file, which can be too big to fit into memory (as is op-mainnet case). This means that we have to call the same `init_*` genesis functions for the same block multiple times in accounts chunks. **This state file is not sorted by account address though**. 

However, at some point`StateRevert::write_to_db` is used  and before  [PR-7784](https://github.com/paradigmxyz/reth/pull/7784/files#diff-43340f91072c31dd8dfcdb7236c267a0563b85c1cf4cdaa710d3a5f62a617c71), **it was expecting either all accounts to be sorted in memory**, or if not, that they should be sorted before calling the function multiple times with all chunks. In the linked PR, there was a change that introduced a  [check + db.upsert](https://github.com/paradigmxyz/reth/pull/7784/files#diff-43340f91072c31dd8dfcdb7236c267a0563b85c1cf4cdaa710d3a5f62a617c71) for when the chunks were not sorted externally.

Why the change?
* `StateRevert::write_to_db` is used during live sync, and that [check + db.upsert](https://github.com/paradigmxyz/reth/pull/7784/files#diff-43340f91072c31dd8dfcdb7236c267a0563b85c1cf4cdaa710d3a5f62a617c71) is not doing anything. It's really only used for this one time command.
* It results in less space used (even sorted inserts are better than unsorted inserts).
* Slighty faster - it's a one time use so whatever.

### Benchmarks
Tested with op mainnet state dump

My machine:
`main` -> `branch`
16min44s -> 11min55s
43.6 GiB -> 32.8GiB

Nice follow-up could be to refactor `crates/node-core/src/init.rs` further. Has way too much in there.

### CODE REVIEW

**Review by mattsse — COMMENTED** [review]
tbh I don't fully understand what this does or why we need etl for init?

> Nice follow-up could be to refactor crates/node-core/src/init.rs further. Has way too much in there.

agree, none of this belongs in node-core

**Comment by joshieDo** [comment]
updated the PR description - lmk if it helps

File: crates/storage/codecs/src/alloy/genesis_account.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+#[main_codec]
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+struct StorageEntry {
+    key: B256,
+    value: B256,
+}
+
+impl Compact for AlloyGenesisAccount {
```
Comment (mattsse) [inline crates/storage/codecs/src/alloy/genesis_account.rs:37]: why do we need this? when are we writing the genesis account to database?

File: crates/node-core/src/init.rs [hunk 2]
```diff
@@ -297,6 +300,7 @@ pub fn insert_genesis_header<DB: Database>(
 pub fn init_from_state_dump<DB: Database>(
```
Comment (mattsse) [inline crates/node-core/src/init.rs:300]: this needs more docs, not obvious what this does

File: crates/storage/codecs/Cargo.toml [hunk 3]
```diff
@@ -17,6 +17,7 @@ reth-codecs-derive = { path = "./derive", default-features = false }
 # eth
 alloy-eips = { workspace = true, optional = true }
 alloy-primitives.workspace = true
+alloy-genesis.workspace = true
```
Comment (mattsse) [inline crates/storage/codecs/Cargo.toml:20]: must be optional and enabled for alloy feature

**Review by mattsse — CHANGES_REQUESTED** [review]
if I understand this correctly, then this uses etl to sort accounts therefor we need compact encoding support for the genesis account object?

File: crates/storage/codecs/src/alloy/genesis_account.rs (hunk 1 above)
Comment (joshieDo) [reply crates/storage/codecs/src/alloy/genesis_account.rs:37]: We don't, but `reth_etl::Collector` requires `V` to impl `Compact`

On early implementations, i was trying to leverage its `serde` with some other serializing lib, but it was failing on deserializing, and was only succeeding with `serde_json` which is too slow.

File: crates/node-core/src/init.rs (hunk 2 above)
Comment (joshieDo) [reply crates/node-core/src/init.rs:305]: Added the following. Does it help?
```rust
/// Reads account state from a [`BufRead`] reader and initializes it at the highest block that can
/// be found on database.
///
/// It's similar to [`init_genesis`] but supports importing state too big to fit in memory, and can
/// be set to the highest block present. One practical usecase is to import OP mainnet state at
/// bedrock transition block.
pub fn init_from_state_dump<DB: Database>(
```

**Review by mattsse — APPROVED** [review]
thanks!

I'll open an issue to move this code somewhere else than node-core

**Comment by emhane** [comment]
this closes https://github.com/paradigmxyz/reth/pull/7898 right?

**Comment by joshieDo** [comment]
oh yeah, forgot to reference that 👍

**Pull request merged**